cmd/object-handlers.go

```diff
@@ -298,50 +316,28 @@ func (api objectAPIHandlers) CopyObjectHandler(w http.ResponseWriter, r *http.Re
 
 	/// maximum Upload size for object in a single CopyObject operation.
 	if isMaxObjectSize(objInfo.Size) {
-		writeErrorResponse(w, r, ErrEntityTooLarge, objectSource)
+		writeErrorResponse(w, r, ErrEntityTooLarge, cpSrcPath)
 		return
 	}
 
-	// Size of object.
-	size := objInfo.Size
-
-	pipeReader, pipeWriter := io.Pipe()
-	go func() {
-		startOffset := int64(0) // Read the whole file.
-		// Get the object.
-		gErr := objectAPI.GetObject(sourceBucket, sourceObject, startOffset, size, pipeWriter)
-		if gErr != nil {
-			errorIf(gErr, "Unable to read an object.")
-			pipeWriter.CloseWithError(gErr)
-			return
-		}
-		pipeWriter.Close() // Close.
-	}()
-
-	// Save other metadata if available.
-	metadata := objInfo.UserDefined
-
-	// Remove the etag from source metadata because if it was uploaded as a multipart object
-	// then its ETag will not be MD5sum of the object.
-	delete(metadata, "md5Sum")
-
-	sha256sum := ""
-
-	objectWLock := globalNSMutex.NewNSLock(bucket, object)
-	objectWLock.Lock()
-	defer objectWLock.Unlock()
+	defaultMeta := objInfo.UserDefined
+	newMetadata := getCpObjMetadataFromHeader(r.Header, defaultMeta)
```

Comment:
newMetadata can be nil here though COPY is the default operation in S3 spec, Also we need to return http 400 when x-amz-metadata-directive is not COPY neither REPLACE